bigquery/google/cloud/bigquery/schema.py

```diff
@@ -256,3 +258,41 @@ def _build_schema_resource(fields):
         Sequence[Dict]: Mappings describing the schema of the supplied fields.
     """
     return [field.to_api_repr() for field in fields]
+
+
+def _to_schema_fields(schema):
+    """Coerce `schema` to a list of schema field instances.
+
+    Args:
+        schema(Union[ \
+               Sequence[:class:`~google.cloud.bigquery.schema.SchemaField`], \
+               Sequence[Mapping[str, str]] \
+        ]):
+            Table schema to convert. If given as a sequence of
+            mappings, their content must be compatible with
+            :meth:`~google.cloud.bigquery.schema.SchemaField.from_api_repr`.
+
+    Returns:
+        Sequence[:class:`~google.cloud.bigquery.schema.SchemaField`]
+
+    Raises:
+        TypeError: If `schema` is not a sequence.
+        ValueError:
+            If any item in the sequence is not a
+            :class:`~google.cloud.bigquery.schema.SchemaField` or a
+            compatible mapping representation of the field.
+    """
+    for field in schema:
+        if not isinstance(field, (SchemaField, collections.Mapping)):
+            raise ValueError(
+                "Schema items must either be fields or compatible "
+                "mapping representations."
+            )
+
+    if schema and not isinstance(schema[0], SchemaField):
+        try:
+            schema = [SchemaField.from_api_repr(field) for field in schema]
+        except Exception as exc:
+            raise ValueError("Invalid field representation: {!r}".format(exc))
+
+    return schema
```

Comment:
This was to preserve the `Table.schema` contract that claims to raise a `ValueError` _"If any item in the sequence is not a SchemaField"_ (to which I added _"or a compatible mapping representation"_).

Since `Table.schema` uses this helper and is exposed to the users, I did not want to alter the former's contract (`to_api_repr()` can also raise a `KeyError`, for instance).

BTW, should we actually allow mixed sequences that can contain both the dicts and SchemaField instances at the same time?

Reply:
I feel it's unlikely that people will need to pass in mixed sequences, but also the implementation feels a bit simpler if we do allow them.

I'm inclined to let `Schema.from_api_repr` throw whatever it's going to throw. I see what you mean about the method contract, but since this is a new feature we can be a flexible in what we throw throw when a sequence of mappings is passed in.

Reply:
If changing the contract is fine, then I'm happy to simplify this part of the code. :+1: